tensilelite/Tensile/KernelWriterAssembly.py

```diff
@@ -10560,6 +10571,279 @@ def writeBiasToGlobal(self, biasDataType, kernel, tP, gwvw, offsetVgpr, tmpSgprR
+    mod.add(label)
+    mod.addSpaceLine()
+    mod.add(VAddU32(vgpr("Tmp"), sgpr("Tmp"), vgpr("Serial")))
+    mod.add(VAndB32(vgpr("Tmp"), wave_size-1, vgpr("Tmp")))
+    mod.add(VLShiftLeftB32(vgpr("Tmp"), 0x2, vgpr("Tmp")))
+    mod.addSpaceLine()
+    mod.add(DSBPermuteB32(vgpr("AmaxOutB"), vgpr("Tmp"), vgpr("AmaxOut")))
+    mod.add(SWaitCnt(lgkmcnt=0))
+    mod.addSpaceLine()
+    # TODO- F16
+    mod.add(VMaxF32(vgpr("AmaxOut"), vgpr("AmaxOut"), vgpr("AmaxOutB")))
+    mod.add(SLShiftLeftB32(sgpr("Tmp"), 1, sgpr("Tmp")))
+    mod.add(SCmpLtU32(sgpr("Tmp"), wave_size))
+    mod.add(SCBranchSCC1(label.getLabelName()))
+    mod.addSpaceLine()
+    return mod
+
+  def amax_inter_wave_reduction(self, kernel) -> Module:
+    wave_size = kernel["WavefrontSize"]
+    numWorkItems = kernel["NumThreads"]
+    amaxOutType = kernel["ProblemType"]["DataTypeAmaxD"]
+    amax_lds_start = kernel["LdsBytesNoAmax"]
+
+    label_wave_inter = Label("wave_inter", 'wave_inter')
+    label_wave_upper = Label("wave_upper", 'wave_upper')
+    label_wave_lower = Label("wave_lower", 'wave_lower')
+    label_wave_empty = Label("wave_empty", 'wave_empty')
+    label_wave_end   = Label("wave_end",   'wave_end')
+
+    mod = Module("inter_wave_reduction")
+    mod.addComment0("inter_wave_reduction")
+
+    mod.add(VLShiftRightB32(vgpr("Widx"), int(log2(wave_size)), vgpr("Serial")))
+    mod.add(SMovB32(sgpr("Offset"), numWorkItems // wave_size))
+    mod.add(label_wave_inter)
+    mod.add(SLShiftRightB32(sgpr("Offset"), 1, sgpr("Offset")))
+    mod.add(SCmpEQU32(sgpr("Offset"), 0))
+    mod.add(SCBranchSCC1(label_wave_end.getLabelName()))
+    mod.add(SLShiftLeftB32(sgpr("Tmp"), 1, sgpr("Offset")))
+    mod.add(VCmpLtU32(sgpr("Tmp+2",2), vgpr("Widx"), sgpr("Tmp")))
+    mod.add(VCmpGEU32(sgpr("Tmp+4",2), vgpr("Widx"), sgpr("Offset")))
+    mod.add(SAndB64("vcc", sgpr("Tmp+2",2), sgpr("Tmp+4",2)))
+    mod.add(SCBranchVCCNZ(label_wave_upper.getLabelName()))
+    mod.add(VCmpLtU32("vcc", vgpr("Widx"), sgpr("Offset")))
+    mod.add(SCBranchVCCNZ(label_wave_lower.getLabelName()))
+    mod.add(SBranch(label_wave_empty.getLabelName()))
+
+    mod.add(label_wave_upper)
+    mod.add(VSubU32(vgpr("Tmp"), vgpr("Widx"), sgpr("Offset")))
+    mod.add(VLShiftLeftB32(vgpr("Tmp"), int(log2(amaxOutType.numBytes())), vgpr("Tmp")))
+
+    # TODO- select inst
+    ds = DSModifiers(offset=amax_lds_start)
+    mod.add(DSStoreB32(vgpr("Tmp"), vgpr("AmaxOut"), ds))
+    mod.add(SWaitCnt(lgkmcnt=0))
+    mod.add(SBarrier())
+    mod.add(SBranch(label_wave_inter.getLabelName()))
+    mod.add(label_wave_lower)
+    mod.add(SBarrier())
+    mod.add(VLShiftLeftB32(vgpr("Tmp"), int(log2(amaxOutType.numBytes())), vgpr("Widx")))
+
+    # TODO- select inst
+    mod.add(DSLoadB32(vgpr("AmaxOutB"), vgpr("Tmp"), ds))
+    mod.add(SWaitCnt(lgkmcnt=0))
+    # TODO- F16
+    mod.add(VMaxF32(vgpr("AmaxOut"), vgpr("AmaxOut"), vgpr("AmaxOutB")))
+    mod.add(SBranch(label_wave_inter.getLabelName()))
```

Comment:
The TODO means now we are considering F32 AmaxD only (due to ticket request). We can make it more generic.